src/nbl/video/IGPUDescriptorSet.cpp

```diff
@@ -5,81 +5,64 @@
 namespace nbl::video
 {
 
-void IGPUDescriptorSet::allocateDescriptors()
+IGPUDescriptorSet::IGPUDescriptorSet(core::smart_refctd_ptr<const IGPUDescriptorSetLayout>&& layout, core::smart_refctd_ptr<IDescriptorPool>&& pool, IDescriptorPool::SDescriptorOffsets&& offsets)
+    : base_t(std::move(layout)), IBackendObject(std::move(core::smart_refctd_ptr<const ILogicalDevice>(pool->getOriginDevice()))), m_version(0ull), m_parentPoolVersion(pool->m_version), m_pool(std::move(pool)), m_descriptorStorageOffsets(std::move(offsets))
 {
-    auto& offsets = m_descriptorStorageOffsets;
-
-    for (uint32_t i = 0u; i < asset::EDT_COUNT; ++i)
+    for (auto i = 0u; i < static_cast<uint32_t>(asset::IDescriptor::E_TYPE::ET_COUNT); ++i)
     {
-        const auto type = static_cast<asset::E_DESCRIPTOR_TYPE>(i);
-        const auto count = getLayout()->getTotalDescriptorCount(type);
-        if (count == 0ull)
+        // There is no descriptor of such type in the set.
+        if (m_descriptorStorageOffsets.data[i] == ~0u)
             continue;
 
-        if (m_pool->m_flags & IDescriptorPool::ECF_FREE_DESCRIPTOR_SET_BIT)
-            offsets.data[type] = m_pool->m_generalAllocators[type].alloc_addr(count, 1u);
-        else
-            offsets.data[type] = m_pool->m_linearAllocators[type].alloc_addr(count, 1u);
+        const auto type = static_cast<asset::IDescriptor::E_TYPE>(i);
 
-        assert((offsets.data[type] < m_pool->m_maxDescriptorCount[i]) && "PANIC: Allocation failed. This shoudn't have happened! Check your descriptor pool.");
+        // Default-construct the core::smart_refctd_ptr<IDescriptor>s because even if the user didn't update the descriptor set with ILogicalDevice::updateDescriptorSet we
+        // won't have uninitialized memory and destruction wouldn't crash in ~IGPUDescriptorSet.
+        std::uninitialized_default_construct_n(m_pool->getDescriptorStorage(type) + m_descriptorStorageOffsets.data[i], m_layout->getTotalDescriptorCount(type));
     }
 
-    const auto mutableSamplerCount = getLayout()->getTotalMutableSamplerCount();
-    if (mutableSamplerCount != 0ull)
-    {
-        if (m_pool->m_flags & IDescriptorPool::ECF_FREE_DESCRIPTOR_SET_BIT)
-            offsets.data[asset::EDT_COUNT] = m_pool->m_generalAllocators[asset::EDT_COUNT].alloc_addr(mutableSamplerCount, 1u);
-        else
-            offsets.data[asset::EDT_COUNT] = m_pool->m_linearAllocators[asset::EDT_COUNT].alloc_addr(mutableSamplerCount, 1u);
-    }
+    const auto mutableSamplerCount = m_layout->getTotalMutableSamplerCount();
+    if (mutableSamplerCount > 0)
+        std::uninitialized_default_construct_n(m_pool->getMutableSamplerStorage() + m_descriptorStorageOffsets.data[static_cast<uint32_t>(asset::IDescriptor::E_TYPE::ET_COUNT)], mutableSamplerCount);
```

Comment:
use `getAllMutableSamplers()` instead of `m_pool->getMutableSamplerStorage() + m_descriptorStorageOffsets.data[static_cast<uint32_t>(asset::IDescriptor::E_TYPE::ET_COUNT)]` ?

Reply:
also you probably don't need to check the count, `_n` should handle the 0 case and prevent from you doing stuff to nullptrs

Reply:
> use `getAllMutableSamplers()` instead of `m_pool->getMutableSamplerStorage() + m_descriptorStorageOffsets.data[static_cast<uint32_t>(asset::IDescriptor::E_TYPE::ET_COUNT)]` ?

Yes, resolved.

Reply:
> also you probably don't need to check the count, `_n` should handle the 0 case and prevent from you doing stuff to nullptrs

The docs doesn't say so, so I left it as is.